File: app/src/main/java/com/arukai/uajpspeak/activity/FragmentDrawer.java
package com.arukai.uajpspeak.activity;

import android.content.Context;
import android.os.Bundle;
import androidx.fragment.app.Fragment;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.appcompat.widget.Toolbar;
import android.view.GestureDetector;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import com.arukai.uajpspeak.R;
import com.arukai.uajpspeak.adapter.NavigationDrawerAdapter;
import com.arukai.uajpspeak.model.NavDrawerItem;

public class FragmentDrawer extends Fragment {

    private static String TAG = FragmentDrawer.class.getSimpleName();

    private RecyclerView recyclerView;
    public ActionBarDrawerToggle mDrawerToggle;
    private DrawerLayout mDrawerLayout;
    public static NavigationDrawerAdapter adapter;
    private View containerView;
    private static String[] titles = null;
    private static int p_counter = 0;
    private FragmentDrawerListener drawerListener;
    public static List<NavDrawerItem> activeData;

    public FragmentDrawer() {
    }

    public void setDrawerListener(FragmentDrawerListener listener) {
        this.drawerListener = listener;
    }

    public List<NavDrawerItem> getData() {
        List<NavDrawerItem> data = new ArrayList<>();

        // preparing navigation drawer items
        for (int i = 0; i < titles.length; i++) {
            NavDrawerItem navItem = new NavDrawerItem();
            navItem.setTitle(titles[i]);
            data.add(navItem);
        }

        p_counter = getActivity().getResources().getStringArray(R.array.greetings).length +
                getActivity().getResources().getStringArray(R.array.signs).length +
                getActivity().getResources().getStringArray(R.array.troubleshooting).length +
                getActivity().getResources().getStringArray(R.array.transportation).length +
                getActivity().getResources().getStringArray(R.array.directions).length +
                getActivity().getResources().getStringArray(R.array.hotel).length +
                getActivity().getResources().getStringArray(R.array.numbers).length +
                getActivity().getResources().getStringArray(R.array.time).length +
                getActivity().getResources().getStringArray(R.array.weekdays).length +
                getActivity().getResources().getStringArray(R.array.months).length +
                getActivity().getResources().getStringArray(R.array.colors).length +
                getActivity().getResources().getStringArray(R.array.common_words).length +
                getActivity().getResources().getStringArray(R.array.restaurant).length +
                getActivity().getResources().getStringArray(R.array.love).length +
                getActivity().getResources().getStringArray(R.array.shopping).length +
                getActivity().getResources().getStringArray(R.array.clothing).length +
                getActivity().getResources().getStringArray(R.array.drugstore).length +
                getActivity().getResources().getStringArray(R.array.driving).length +
                getActivity().getResources().getStringArray(R.array.bank).length;

        data.get(MainActivity.current_position).setSelected(true);
        return data;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // drawer labels
        titles = getActivity().getResources().getStringArray(R.array.nav_drawer_labels);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        // Inflating view layout
        View layout = inflater.inflate(R.layout.fragment_navigation_drawer, container, false);
        recyclerView = layout.findViewById(R.id.drawerList);
        activeData = getData();
        adapter = new NavigationDrawerAdapter(getActivity(), activeData);
        recyclerView.setAdapter(adapter);
        recyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
        recyclerView.addOnItemTouchListener(new RecyclerTouchListener(getActivity(), recyclerView, new ClickListener() {
            @Override
            public void onClick(View view, int position) {
                drawerListener.onDrawerItemSelected(view, position);
                mDrawerLayout.closeDrawer(containerView);
            }

            @Override
            public void onLongClick(View view, int position) {

            }
        }));
        layout.setOnClickListener(null);

        TextView phrases_count = layout.findViewById(R.id.phrasesCount);
        phrases_count.setText(p_counter+"フレーズ");
        return layout;
    }

    public void setUp(int fragmentId, DrawerLayout drawerLayout, final Toolbar toolbar) {
        containerView = getActivity().findViewById(fragmentId);
        mDrawerLayout = drawerLayout;
        mDrawerToggle = new ActionBarDrawerToggle(getActivity(), drawerLayout, toolbar, R.string.drawer_open, R.string.drawer_close) {
            @Override
            public void onDrawerOpened(View drawerView) {

                if(MainActivity.isSearchOpened) {
                    ActionBar action = ((MainActivity)getActivity()).getSupportActionBar();
                    action.setDisplayShowCustomEnabled(false); //disable a custom view inside the actionbar
                    action.setDisplayShowTitleEnabled(true); //show the title in the action bar

                    //hides the keyboard
                    final InputMethodManager imm = (InputMethodManager) getActivity().getSystemService(Context.INPUT_METHOD_SERVICE);
                    imm.hideSoftInputFromWindow(getView().getWindowToken(), 0);

                    MainActivity.isSearchOpened = false;
                }

                super.onDrawerOpened(drawerView);
                getActivity().invalidateOptionsMenu();
            }

            @Override
            public void onDrawerClosed(View drawerView) {
                super.onDrawerClosed(drawerView);
                getActivity().invalidateOptionsMenu();
            }

            @Override
            public void onDrawerSlide(View drawerView, float slideOffset) {
                super.onDrawerSlide(drawerView, slideOffset);
                toolbar.setAlpha(1 - slideOffset / 2);
            }
        };

        mDrawerLayout.setDrawerListener(mDrawerToggle);
        mDrawerLayout.post(new Runnable() {
            @Override
            public void run() {
                mDrawerToggle.syncState();
            }
        });
    }

    public interface ClickListener {
        void onClick(View view, int position);
        void onLongClick(View view, int position);
    }

    static class RecyclerTouchListener implements RecyclerView.OnItemTouchListener {

        private GestureDetector gestureDetector;
        private ClickListener clickListener;

        public RecyclerTouchListener(Context context, final RecyclerView recyclerView, final ClickListener clickListener) {
            this.clickListener = clickListener;
            gestureDetector = new GestureDetector(context, new GestureDetector.SimpleOnGestureListener() {
                @Override
                public boolean onSingleTapUp(MotionEvent e) {
                    return true;
                }

                @Override
                public void onLongPress(MotionEvent e) {
                    View child = recyclerView.findChildViewUnder(e.getX(), e.getY());
                    if (child != null && clickListener != null) {
                        clickListener.onLongClick(child, recyclerView.getChildPosition(child));
                    }
                }
            });
        }

        @Override
        public boolean onInterceptTouchEvent(RecyclerView rv, MotionEvent e) {

            View child = rv.findChildViewUnder(e.getX(), e.getY());
            if (child != null && clickListener != null && gestureDetector.onTouchEvent(e)) {
                clickListener.onClick(child, rv.getChildPosition(child));
            }
            return false;
        }

        @Override
        public void onTouchEvent(RecyclerView rv, MotionEvent e) {
        }

        @Override
        public void onRequestDisallowInterceptTouchEvent(boolean disallowIntercept) {

        }
    }

    public interface FragmentDrawerListener {
        void onDrawerItemSelected(View view, int position);
    }

    public void setDrawerState(boolean enabled) {
        int lockMode = enabled ? DrawerLayout.LOCK_MODE_UNLOCKED :
                DrawerLayout.LOCK_MODE_LOCKED_CLOSED;
        mDrawerLayout.setDrawerLockMode(lockMode);
        mDrawerToggle.setDrawerIndicatorEnabled(enabled);
    }
}

File: app/src/main/java/com/arukai/uajpspeak/activity/FragmentDrawer.kt
package com.arukai.uajpspeak.activity

import android.content.Context
import android.os.Bundle
import android.view.GestureDetector
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import android.widget.TextView
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.widget.Toolbar
import androidx.drawerlayout.widget.DrawerLayout
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.arukai.uajpspeak.R
import com.arukai.uajpspeak.adapter.NavigationDrawerAdapter
import com.arukai.uajpspeak.model.NavDrawerItem

class FragmentDrawer : Fragment() {

    private lateinit var recyclerView: RecyclerView
    lateinit var mDrawerToggle: ActionBarDrawerToggle
    private lateinit var mDrawerLayout: DrawerLayout
    private lateinit var containerView: View
    private var drawerListener: FragmentDrawerListener? = null

    companion object {
        var adapter: NavigationDrawerAdapter? = null
        var activeData: MutableList<NavDrawerItem>? = null
        private var titles: Array<String>? = null
        private var pCounter = 0
    }

    interface FragmentDrawerListener {
        fun onDrawerItemSelected(view: View, position: Int)
    }

    interface ClickListener {
        fun onClick(view: View, position: Int)
        fun onLongClick(view: View, position: Int)
    }

    private fun getData(): MutableList<NavDrawerItem> {
        val data = ArrayList<NavDrawerItem>()

        titles?.forEach { title ->
            val navItem = NavDrawerItem()
            navItem.title = title
            data.add(navItem)
        }

        pCounter = resources.getStringArray(R.array.greetings).size +
                resources.getStringArray(R.array.signs).size +
                resources.getStringArray(R.array.troubleshooting).size +
                resources.getStringArray(R.array.transportation).size +
                resources.getStringArray(R.array.directions).size +
                resources.getStringArray(R.array.hotel).size +
                resources.getStringArray(R.array.numbers).size +
                resources.getStringArray(R.array.time).size +
                resources.getStringArray(R.array.weekdays).size +
                resources.getStringArray(R.array.months).size +
                resources.getStringArray(R.array.colors).size +
                resources.getStringArray(R.array.common_words).size +
                resources.getStringArray(R.array.restaurant).size +
                resources.getStringArray(R.array.love).size +
                resources.getStringArray(R.array.shopping).size +
                resources.getStringArray(R.array.clothing).size +
                resources.getStringArray(R.array.drugstore).size +
                resources.getStringArray(R.array.driving).size +
                resources.getStringArray(R.array.bank).size

        data[MainActivity.current_position].isSelected = true
        return data
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        titles = resources.getStringArray(R.array.nav_drawer_labels)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val layout = inflater.inflate(R.layout.fragment_navigation_drawer, container, false)
        recyclerView = layout.findViewById(R.id.drawerList)
        activeData = getData()
        adapter = NavigationDrawerAdapter(requireActivity(), activeData!!)
        recyclerView.adapter = adapter
        recyclerView.layoutManager = LinearLayoutManager(activity)
        recyclerView.addOnItemTouchListener(
            RecyclerTouchListener(
                requireActivity(),
                recyclerView,
                object : ClickListener {
                    override fun onClick(view: View, position: Int) {
                        drawerListener?.onDrawerItemSelected(view, position)
                        mDrawerLayout.closeDrawer(containerView)
                    }

                    override fun onLongClick(view: View, position: Int) {
                    }
                })
        )
        layout.setOnClickListener(null)

        // Update the NavigationView header (NavigationView hosts the header layout `nav_header.xml`)
        // Find the NavigationView via the hosting Activity and update its header banner and phrasesCount
        val navView = activity?.findViewById<com.google.android.material.navigation.NavigationView>(R.id.navigation_view)
        navView?.getHeaderView(0)?.let { header ->
            val banner = header.findViewById<android.widget.ImageView>(R.id.banner)
            val lang = com.arukai.uajpspeak.util.LocaleHelper.getSavedLanguage(requireContext())
            val sourceFlagRes = when (lang) {
                "en" -> R.drawable.uk
                "de" -> R.drawable.de
                "es" -> R.drawable.es
                "fr" -> R.drawable.fr
                "ja" -> R.drawable.jp
                else -> R.drawable.uk
            }
            banner?.setImageResource(sourceFlagRes)

            val phrasesCountView = header.findViewById<TextView>(R.id.phrasesCount)
            phrasesCountView?.text = getString(R.string.phrases_counter, pCounter)
        }
        return layout
    }

    fun setUp(fragmentId: Int, drawerLayout: DrawerLayout, toolbar: Toolbar) {
        containerView = requireActivity().findViewById(fragmentId)
        mDrawerLayout = drawerLayout
        mDrawerToggle = object : ActionBarDrawerToggle(
            activity,
            drawerLayout,
            toolbar,
            R.string.drawer_open,
            R.string.drawer_close
        ) {
            override fun onDrawerOpened(drawerView: View) {
                if (MainActivity.isSearchOpened) {
                    val action = (activity as MainActivity).supportActionBar
                    action?.setDisplayShowCustomEnabled(false)
                    action?.setDisplayShowTitleEnabled(true)

                    val imm = activity?.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
                    view?.let { imm.hideSoftInputFromWindow(it.windowToken, 0) }

                    MainActivity.isSearchOpened = false
                }

                super.onDrawerOpened(drawerView)
                activity?.invalidateOptionsMenu()
            }

            override fun onDrawerClosed(drawerView: View) {
                super.onDrawerClosed(drawerView)
                activity?.invalidateOptionsMenu()
            }

            override fun onDrawerSlide(drawerView: View, slideOffset: Float) {
                super.onDrawerSlide(drawerView, slideOffset)
                toolbar.alpha = 1 - slideOffset / 2
            }
        }

        mDrawerLayout.addDrawerListener(mDrawerToggle)
        mDrawerLayout.post {
            mDrawerToggle.syncState()
        }
    }

    class RecyclerTouchListener(
        context: Context,
        private val recyclerView: RecyclerView,
        private val clickListener: ClickListener
    ) : RecyclerView.OnItemTouchListener {

        private val gestureDetector: GestureDetector = GestureDetector(context, object : GestureDetector.SimpleOnGestureListener() {
            override fun onSingleTapUp(e: MotionEvent): Boolean {
                return true
            }

            override fun onLongPress(e: MotionEvent) {
                val child = recyclerView.findChildViewUnder(e.x, e.y)
                if (child != null) {
                    clickListener.onLongClick(child, recyclerView.getChildAdapterPosition(child))
                }
            }
        })

        override fun onInterceptTouchEvent(rv: RecyclerView, e: MotionEvent): Boolean {
            val child = rv.findChildViewUnder(e.x, e.y)
            if (child != null && gestureDetector.onTouchEvent(e)) {
                clickListener.onClick(child, rv.getChildAdapterPosition(child))
            }
            return false
        }

        override fun onTouchEvent(rv: RecyclerView, e: MotionEvent) {
        }

        override fun onRequestDisallowInterceptTouchEvent(disallowIntercept: Boolean) {
        }
    }
}
